src/Bicep.Core/Analyzers/Linter/Rules/BCPL1060.cs

```diff
@@ -0,0 +1,186 @@
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Text;
+
+namespace Bicep.Core.Analyzers.Linter.Rules
+{
+    public sealed class BCPL1060 : LinterRule
+    {
+        public BCPL1060() : base(
+            code: "BCPL1060",
+            ruleName: "Dynamic variable used concat",
+            description: "Dynamic variable should not use concat - string interpolation should be used.",
+            docUri: "https://bicep/linter/rules/BCPL1060") // TODO: setup up doc pages
+        { }
+
+        private CodeReplacement GetCodeReplacement(TextSpan span)
+            => new CodeReplacement(span, "BCPL1060 - this is the new code");
+
+        override public IEnumerable<IBicepAnalyzerDiagnostic> Analyze(SemanticModel model)
+        {
+            var funcSymbols = model.Root.Declarations.OfType<VariableSymbol>()
+                                    .Select(sym => sym.Value)
+                                    .OfType<FunctionCallSyntax>();
+
+            foreach (var func in funcSymbols.Where(f => f.Name.IdentifierName.Equals("concat", System.StringComparison.OrdinalIgnoreCase)))
+            {
+                if (CreateFix(func) is CodeFix fix)
+                {
+                    yield return CreateFixableDiagnosticForSpan(func.Span, fix);
+                }
+            }
+        }
+
+        private CodeFix? CreateFix(FunctionCallSyntax func)
+        {
+            if (GetCodeReplacement(func) is CodeReplacement cr)
+            {
+                return new CodeFix($"Use string interpolation: {cr.Text}", true, cr);
+            }
+            return null;
+        }
+
+        private CodeReplacement? GetCodeReplacement(FunctionCallSyntax func)
+        {
+            if (RewriteConcatToInterpolate(func) is StringSyntax newSyntax)
+            {
+                return CodeReplacement.FromSyntax(func.Span, newSyntax);
+            }
+            return null;
+        }
+
+        private StringSyntax? RewriteConcatToInterpolate(FunctionCallSyntax func)
+        {
+            var rewrite = CallbackConvertorRewriter<FunctionCallSyntax, StringSyntax>.Rewrite(func, RewriteConcatCallback);
+            return rewrite;
+        }
+
+        private StringSyntax RewriteConcatCallback(FunctionCallSyntax syntax)
+        {
+            var tokens = new List<Token>();
+            var expressions = new List<SyntaxBase>();
+            var segments = new List<string>();
+
+            var flattened = SyntaxFactory.FlattenStringOperations(syntax);
+            if (flattened is FunctionCallSyntax concatSyntax)
+            {
+                return CreateStringInterpolation(concatSyntax.Arguments.Select(a => a.Expression).ToImmutableArray());
+            }
+            else if (flattened is StringSyntax stringSyntax)
+            {
+                return stringSyntax;
+            }
+
+            // TODO:  What is the correct way to handle a failed codefix?
```

Comment:
>TODO [](http://example.com/codeflow?start=15&length=4)

Does LSP allow us to signal failure to apply the codefix? @shenglol do you remember?

Reply:
Yes, I think we can catch the exception in the code action handler and return a [Command](https://microsoft.github.io/language-server-protocol/specifications/specification-current/#command) to let VSCode show an error message.